### ISSUE TITLE
Drop support for pyOpenSSL < 0.10

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5014]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5014|
|Type|task|
|Created|2011-04-03 15:49:03Z|
|Branch|https://github.com/twisted/twisted/tree/old-as-hell-openssl-5014|

In [#4854](https://github.com/twisted/twisted/issues/4854), we added support for using the memory bio features of pyOpenSSL >= 0.10 to implement `IReactorSSL` in all reactors.

In [[#4974](https://github.com/twisted/twisted/issues/4974)](https://github.com/twisted/twisted/issues/4974), we will deprecate the support based on pyOpenSSL < 0.10.

After that deprecation has been in place for some time, we should remove support for working with pyOpenSSL < 0.10.

If #4974 is resolved for 11.1, then I think 12.1 would be the right time to do this.  When we do, we can get rid of all of `twisted/internet/_oldtls.py` and the conditionals elsewhere which import that module when pyOpenSSL < 0.10 is in use.

Attachments:

* [t5014.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ce/5cea478733fd836f7011cad7ebcab19ffc029d96/b95b73d0c54a0b1866c3b2102bd1959c79ec1d3b.diff) (18725 bytes) - added by Alex on 2014-03-19 21:57:48Z - 
* [t5014.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ce/5cea478733fd836f7011cad7ebcab19ffc029d96/eebd75330860c9a066d33c4417611c41839a8227.diff) (18782 bytes) - added by Alex on 2014-03-19 22:01:37Z - Same patch with a better topfile.
* [t5014.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ce/5cea478733fd836f7011cad7ebcab19ffc029d96/8a90b1327cf5eb76c1f2561d1eb158b4f97d1623.diff) (18785 bytes) - added by Alex on 2014-03-20 14:20:46Z - Fixed!
* [t5014.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ce/5cea478733fd836f7011cad7ebcab19ffc029d96/64fc7774f1c04f01e2f527bc181b162fee0aa5c8.diff) (19491 bytes) - added by Alex on 2014-03-23 16:08:09Z - Updated INSTALL file.

<details><summary>Searchable metadata</summary>

```
trac-id__5014 5014
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_old_as_hell_openssl_5014 branches/old-as-hell-openssl-5014
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1301845743000000 1301845743000000
changetime__1395702004000000 1395702004000000
version__None None
owner__Alex Alex
cc__thijs cc__davidsarah cc__hynek cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [31537]) Merge protocol-ssl-4854-6 

Author: exarkun
Reviewer: itamar, glyph, thijs
Fixes: [#4854](https://github.com/twisted/twisted/issues/4854)
Refs: [#4974](https://github.com/twisted/twisted/issues/4974)
Refs: [#5014](https://github.com/twisted/twisted/issues/5014)
Refs: [#4455](https://github.com/twisted/twisted/issues/4455)

Add an implementation of `IReactorSSL` and `ITLSTransport` which uses the memory
BIO APIs present in pyOpenSSL 0.10 and newer.  This implementation will be preferred
by all reactors if the pyOpenSSL dependency is satisfied, otherwise the old
implementation will still be used.

This appears to have slightly better performance than the old implementation and
should avoid bugs like [#4455](https://github.com/twisted/twisted/issues/4455).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

Removing 12.1 milestone, since [#4974](https://github.com/twisted/twisted/issues/4974) is still outstanding.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

The topfile is missing a 'R' from the start of the line. 

Also, what do you think about starting the topfile with 'twisted.internet.tcp' to identify the place where changes were made.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|Alex commented|
|-|-|

What do you mean "start the topfile with 'twisted.internet.tcp'", I was told the topfile should only mention the user visible changes, and not the internal details (i.e. oldtls).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

True. No need for 'twisted.internet.tcp'
There are still some typos in the message

Maybe used something based on example from: http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

```
Support for versions of pyOpenSSL older than 0.10 has been removed. Affected users should upgrade pyOpenSSL. 
```

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

Thanks for the fix. Everything is fine by me and can be merged.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to Alex|
|-|-|

The top-level INSTALL file still says that older versions of pyOpenSSL are supported.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

(In [41985]) Branching to old-as-hell-openssl-5014.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_16"></a>|@radix commented|
|-|-|

Looks good. I think it would be cool to say ", but version 0.14 is recommended for enhanced security." in the INSTALL and news files, but it's not imperative.

There are some trivial pyflakes and twistedchecker errors. It looks like behavioral tests are passing on all the builders, so please fix those and merge assuming all the rest of the builders are green.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_17"></a>|@zooko commented|
|-|-|

Replying to [radix](#note_16):
> Looks good. I think it would be cool to say ", but version 0.14 is recommended for enhanced security." in the INSTALL and news files, but it's not imperative.

Could you be more specific about the "enhanced security"? For the Tahoe-LAFS project we're currently planning to pin our dependency on pyOpenSSL to < 0.14 in order to deal with https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2193 , and I'd like for us (and our users) to know exactly what security risks we might incur by doing that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

Replying to [zooko](#note_17):
> Replying to [radix](#note_16):
> > Looks good. I think it would be cool to say ", but version 0.14 is recommended for enhanced security." in the INSTALL and news files, but it's not imperative.
> 
> Could you be more specific about the "enhanced security"? For the Tahoe-LAFS project we're currently planning to pin our dependency on pyOpenSSL to < 0.14 in order to deal with https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2193 , and I'd like for us (and our users) to know exactly what security risks we might incur by doing that.

One thing is [ticket:6586 ECDHE key exchange]. Another is that instead of a bunch of custom-written C code, pyOpenSSL is now almost entirely automatically generated from a verified set of static metadata, eliminating an entire class of potential bugs.

Pinning this dependency would IMHO be a serious mistake, but I'll discuss that with you on the relevant bug, not here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

It would be good for INSTALL to mention the capabilities that 0.14 allows, but I think the NEWS file doesn't need to talk about "enhanced security"; it just describes the change being made here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

Good to land, I'll do so.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set status to `closed`|
|-|-|

(In [41990]) Merge old-as-hell-openssl-5014: Drop support for older pyOpenSSL.

Author: Alex

Reviewer: exarkun, glyph

Fixes: [#5014](https://github.com/twisted/twisted/issues/5014)

Remove support for pyOpenSSL older than 0.10
